Source/Diagnostics/ReducedDiags/DifferentialLuminosity.cpp

```diff
@@ -0,0 +1,287 @@
+
+    // read bin parameters
+    int bin_num = 0;
+    amrex::Real bin_max = 0.0_rt, bin_min = 0.0_rt;
+    utils::parser::getWithParser(pp_rd_name, "bin_number", bin_num);
+    utils::parser::getWithParser(pp_rd_name, "bin_max",    bin_max);
+    utils::parser::getWithParser(pp_rd_name, "bin_min",    bin_min);
+    m_bin_num = bin_num;
+    m_bin_max = bin_max;
+    m_bin_min = bin_min;
+    m_bin_size = (bin_max - bin_min) / bin_num;
+
+    // resize and zero-out data array
+    m_data.resize(m_bin_num,0.0_rt);
+
+    if (amrex::ParallelDescriptor::IOProcessor())
+    {
+        if ( m_write_header )
+        {
+            // open file
+            std::ofstream ofs;
+            ofs.open(m_path + m_rd_name + "." + m_extension,
+                std::ofstream::out | std::ofstream::app);
+            // write header row
+            int off = 0;
+            ofs << "#";
+            ofs << "[" << off++ << "]step()";
+            ofs << m_sep;
+            ofs << "[" << off++ << "]time(s)";
+            for (int i = 0; i < m_bin_num; ++i)
+            {
+                ofs << m_sep;
+                ofs << "[" << off++ << "]";
+                const Real b = m_bin_min + m_bin_size*(Real(i)+0.5_rt);
+                ofs << "bin" << 1+i << "=" << b << "(J)";
+            }
+            ofs << std::endl;
+            // close file
+            ofs.close();
+        }
+    }
+}
+
+void DifferentialLuminosity::ComputeDiags (int step)
+{
+#if defined(WARPX_DIM_RZ)
+    amrex::ignore_unused(step);
+#else
+    WARPX_PROFILE("DifferentialLuminosity::ComputeDiags");
+
+    // Judge if the diags should be done
+    if (!m_intervals.contains(step+1)) { return; }
+
+    Real c2 = PhysConst::c*PhysConst::c;
+
+    // get a reference to WarpX instance
+    auto& warpx = WarpX::GetInstance();
+    Real dt = warpx.getdt(0);
+    // get cell volume
+    Geometry const & geom = warpx.Geom(0);
+    const Real dV = AMREX_D_TERM(geom.CellSize(0), *geom.CellSize(1), *geom.CellSize(2));
+
+    // declare local variables
+    auto const num_bins = m_bin_num;
+    Real const bin_min  = m_bin_min;
+    Real const bin_size = m_bin_size;
+
+    // get MultiParticleContainer class object
+    const MultiParticleContainer& mypc = warpx.GetPartContainer();
+
+    auto& species_1 = mypc.GetParticleContainerFromName(m_beam_name[0]);
+    auto& species_2 = mypc.GetParticleContainerFromName(m_beam_name[1]);
+
+    const ParticleReal m1 = species_1.getMass();
+    const ParticleReal m2 = species_2.getMass();
+
+    // zero-out old data on the host
+    //std::fill(m_data.begin(), m_data.end(), amrex::Real(0.0));
+    //amrex::Gpu::DeviceVector< amrex::Real > d_data( m_data.size(), 0.0 );
+    //amrex::Real* const AMREX_RESTRICT dptr_data = d_data.dataPtr();
+
+    amrex::Gpu::DeviceVector< amrex::Real > d_data( m_data.size() );
+    amrex::Gpu::copy(amrex::Gpu::hostToDevice,
+        m_data.begin(), m_data.end(), d_data.begin());
```

Comment:
We should move this to the constructor

Reply:
Uhm, why? This copies `m_data` to the GPU, without zeroing it out, right?
The initialization to zero is only in the constructor, where:
`m_data.resize(m_bin_num,0.0_rt);`

Reply:
This is what I meant:
https://github.com/ECP-WarpX/WarpX/pull/5161/commits/1402ba2124da0cdfb9068488658953e2069e1aa1